src/Microsoft.AspNetCore.ResponseCaching/Internal/ResponseCachePolicyProvider.cs

```diff
@@ -57,25 +58,51 @@ public virtual bool IsRequestCacheable(ResponseCacheContext context)
 
         public virtual bool IsResponseCacheable(ResponseCacheContext context)
         {
+            var responseCacheControlHeader = context.HttpContext.Response.Headers[HeaderNames.CacheControl];
+
             // Only cache pages explicitly marked with public
-            if (!context.ResponseCacheControlHeaderValue.Public)
+            var isPublic = false;
+            foreach (var header in responseCacheControlHeader)
+            {
+                if (header.IndexOf(CacheControlValues.PublicString, StringComparison.OrdinalIgnoreCase) != -1)
+                {
+                    isPublic = true;
+                    break;
+                }
+            }
+            if (!isPublic)
             {
                 context.Logger.LogResponseWithoutPublicNotCacheable();
                 return false;
             }
 
             // Check no-store
-            if (context.RequestCacheControlHeaderValue.NoStore || context.ResponseCacheControlHeaderValue.NoStore)
+            foreach (var header in context.HttpContext.Request.Headers[HeaderNames.CacheControl])
```

Comment:
This could be extracted in a helper as well. `ParsingHelper.HeaderContains(string)`